p2panda-sync-next/src/traits.rs

```diff
@@ -20,6 +22,33 @@ pub trait Protocol {
     ) -> impl Future<Output = Result<Self::Output, Self::Error>>;
 }
 
+/// Interface for managing sync sessions and consuming events they emit.
+pub trait SyncManager<T> {
+    type Protocol: Protocol + Configurable<Config = SyncSessionConfig<T>>;
+    type Event;
+    type Error: Debug;
+
+    /// Instantiate a new sync session.
+    fn session(&mut self, session_id: u64) -> Self::Protocol;
+
+    /// Retrieve a send handle to an already existing sync session.
+    fn session_handle(
+        &self,
+        session_id: u64,
+    ) -> Option<impl Sink<ToSync, Error = Self::Error> + 'static>;
+
+    /// Drive the manager to process and return events emitted from all running sync sessions.
+    fn next_event(&mut self) -> impl Future<Output = Result<Option<Self::Event>, Self::Error>>;
+}
+
+/// Generic interface for configurable types.
+pub trait Configurable {
+    type Error;
+    type Config;
+
+    fn configure(&mut self, config: &Self::Config) -> Result<(), Self::Error>;
+}
```

Comment:
This feels a bit overly generic while you maybe had a concrete use-case in mind (I think the `Protocol + Configurable<Config = SyncSessionConfig<T>>;` hints on that), why not have the configuration on the concrete implementation instead?

Reply:
As we have it now any manager implementation used in p2panda-net needs to produce sessions which can be configured with the concrete `SyncSessionConfig` type exported from `p2panda-sync`.

Reply:
Maybe as a "guideline", I think there is two reasons to introduce generics:

1. Nice separation of systems to integrate this in, for example actors
2. Different types being used in different places

I don't see right now how either of this is needed for configurations (as you are using only one concrete configuration type and it will not change in the context of actors either).

Reply:
Let's make it generic when the need arises and keep it "simple" for now?